kafka-streams/src/main/java/io/micronaut/configuration/kafka/streams/health/KafkaStreamsHealth.java

```diff
@@ -0,0 +1,211 @@
+ * You may obtain a copy of the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package io.micronaut.configuration.kafka.streams.health;
+
+import io.micronaut.configuration.kafka.streams.ConfiguredStreamBuilder;
+import io.micronaut.configuration.kafka.streams.KafkaStreamsConfiguration;
+import io.micronaut.configuration.kafka.streams.KafkaStreamsFactory;
+import io.micronaut.context.annotation.Requires;
+import io.micronaut.health.HealthStatus;
+import io.micronaut.management.health.aggregator.HealthAggregator;
+import io.micronaut.management.health.indicator.HealthIndicator;
+import io.micronaut.management.health.indicator.HealthResult;
+import io.reactivex.Flowable;
+import org.apache.kafka.streams.KafkaStreams;
+import org.apache.kafka.streams.StreamsConfig;
+import org.apache.kafka.streams.processor.TaskMetadata;
+import org.apache.kafka.streams.processor.ThreadMetadata;
+import org.reactivestreams.Publisher;
+
+import javax.inject.Singleton;
+import java.util.HashMap;
+import java.util.List;
+import java.util.Map;
+import java.util.Optional;
+import java.util.Properties;
+import java.util.Set;
+import java.util.stream.Collectors;
+
+/**
+ * A {@link HealthIndicator} for Kafka Streams.
+ *
+ * @author Christian Oestreich
+ * @since 2.0.1
+ */
+@Singleton
+@Requires(classes = HealthIndicator.class)
+@Requires(property = KafkaStreamsConfiguration.PREFIX + ".health.enabled", value = "true", defaultValue = "true")
+public class KafkaStreamsHealth implements HealthIndicator {
+
+    private static final String NAME = "kafkaStreams";
+
+    private final KafkaStreamsFactory kafkaStreamsFactory;
+    private final HealthAggregator<?> healthAggregator;
+
+    /**
+     * Constructor for the health check.
+     *
+     * @param kafkaStreamsFactory The stream factory to get streams from
+     * @param healthAggregator    Health aggregator
+     */
+    public KafkaStreamsHealth(final KafkaStreamsFactory kafkaStreamsFactory, final HealthAggregator<?> healthAggregator) {
+        this.kafkaStreamsFactory = kafkaStreamsFactory;
+        this.healthAggregator = healthAggregator;
+    }
+
+    /**
+     * Get the health result of the streams.  Will attempt to interrogate
+     * details of each stream as well.  The application.id will be used
+     * for each configured stream as the primary health check name.
+     *
+     * @return Health Result Aggregate
+     */
+    @Override
+    public Publisher<HealthResult> getResult() {
+        Flowable<HealthResult> kafkaStreamHealth = Flowable.fromIterable(kafkaStreamsFactory.getStreams().keySet())
```

Comment:
Do we want one health aggregated health result or many?  if we do `fromIterable`, we will get many.  You can flatten them into one result if that is what actually is desired.

Reply:
We want a high level single status that comprises of on status per kafkastream once the aggregator with name at the end

Reply:
If that's the case, then you will want to do `Flowable.just` with the iterable as the parameter, and then you can have a list of generated details pertaining to the individual `HealthResult`.  I did something similar for hystrix as seen here:

```java
@Singleton
--
  | @AllArgsConstructor
  | @Requires(beans = HystrixInterceptor.class)
  | public class HystrixHealthIndicator implements HealthIndicator {
  |  
  | private static final String HEALTH_RESULT_NAME = "hystrixCircuitBreaker";
  | private static final String CIRCUIT_OPEN = "CIRCUIT_OPEN";
  | private static final String CIRCUIT_CLOSED = "CIRCUIT_CLOSED";
  |  
  | /** {@inheritDoc} */
  | @Override
  | public Publisher<HealthResult> getResult() {
  | return Flowable.fromIterable(HystrixCommandMetrics.getInstances())
  | .flatMapMaybe(HystrixHealthIndicator::aggregateCircuitBreakers)
  | .defaultIfEmpty(HealthResult.builder(HEALTH_RESULT_NAME, HealthStatus.UP))
  | .map(HealthResult.Builder::build);
  | }
  |  
  | /**
  | * Aggregate circuit breakers into health results
  | *
  | * @param hystrixCommandMetrics the hystrix command metrics
  | * @return the aggregated circuit breaker health results
  | */
  | private static Maybe<HealthResult.Builder> aggregateCircuitBreakers(final HystrixCommandMetrics hystrixCommandMetrics) {
  | return Single.fromCallable(hystrixCommandMetrics::getCommandKey)
  | .map(HystrixCircuitBreaker.Factory::getInstance)
  | .map(Optional::ofNullable)
  | .filter(Optional::isPresent)
  | .map(Optional::get)
  | .map(hystrixCircuitBreaker -> HealthResult.builder(String.format("%s-%s", HEALTH_RESULT_NAME, hystrixCommandMetrics.getCommandKey().name()))
  | .status(deriveCircuitBreakerHealthStatus(hystrixCommandMetrics, hystrixCircuitBreaker))
  | .details(hystrixCommandMetrics.getHealthCounts()))
  | .defaultIfEmpty(HealthResult.builder(HEALTH_RESULT_NAME, HealthStatus.UP))
  | .onErrorReturn(e -> HealthResult.builder(HEALTH_RESULT_NAME, HealthStatus.UNKNOWN).exception(e));
  | }
  |  
  | /**
  | * Derive the {@link HealthStatus} for the circuit based on whether or not its open
  | *
  | * @param hystrixCommandMetrics the hystrix command metrics
  | * @param hystrixCircuitBreaker the hystrix circuit breaker
  | * @return the health status
  | */
  | private static HealthStatus deriveCircuitBreakerHealthStatus(final HystrixCommandMetrics hystrixCommandMetrics, final HystrixCircuitBreaker hystrixCircuitBreaker) {
  | if (hystrixCircuitBreaker.isOpen()) {
  | return new HealthStatus(CIRCUIT_OPEN, null, false, hystrixCommandMetrics.getHealthCounts().getErrorPercentage() * 100);
  | }
  | return new HealthStatus(CIRCUIT_CLOSED, null, true, null);
  | }
  |  
  | }


```

Reply:
Yeah, I think you are all set here then.  Overlooked the `aggregate` at the bottom.

Reply:
👍